File: OOP/7_Function_Overriding.cpp
#include <bits/stdc++.h>
using namespace std;

// Base class
class Animal {
public:
    // Virtual function for overriding
    virtual void sound() {
        cout << "Some generic animal sound" << endl;
    }
};

// Derived class Dog
class Dog : public Animal {
public:
    // Overriding the sound method
    void sound() override {
        cout << "Bark" << endl;
    }
};

// Derived class Cat
class Cat : public Animal {
public:
    // Overriding the sound method
    void sound() override {
        cout << "Meow" << endl;
    }
};

int main() {
    // Creating objects of different classes
    Animal* animalPtr;  // Pointer of type Animal

    Dog dog;
    Cat cat;

    // Using polymorphism to call overridden methods
    animalPtr = &dog;
    animalPtr->sound();  // Output: Bark (calls Dog's sound method)

    animalPtr = &cat;
    animalPtr->sound();  // Output: Meow (calls Cat's sound method)

    return 0;
}

File: OOP/7_Function_Overriding.py




# Runtime Polymorphism -> Dynamic Binding | Overriding | Late Binding

# CompileTime Polymorphism-> Static Binding | Overloading | Early Binding  



class Animal:
    def sound(self):
        return "Animal Sound "

class Dog(Animal):
    def sound(self):  #overriding parent class method
        return "Bark"

class Cat(Animal):
    def sound(self):  #overriding parent class method
        return "Meow"


# Polymorphism: Different behaviors depending on the actual object type
def make_sound(animal):
    print (animal.sound())


# print(Dog().sound())
dog = Dog()
cat = Cat()

make_sound(Dog())
make_sound(dog)



